### PR TITLE
Add Waitlist local package

### PR DESCRIPTION
**Pull request by ayoy**
Task/Issue URL: https://app.asana.com/0/0/1203937201668213/f

**Description**:
Extract existing waitlist code into a local package to simplify creating waitlists.

**Steps to test this PR**:
See Asana task for detailed testing steps.

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

### CODE REVIEW

File: DuckDuckGo/FailedAssertionView.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import SwiftUI
+
+struct FailedAssertionView: View {
```
Comment (samsymons) [inline DuckDuckGo/FailedAssertionView.swift:22]: Nice 👍

File: LocalPackages/Waitlist/Tests/WaitlistTests/WaitlistViewModelTests.swift [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+import WaitlistMocks
+@testable import Waitlist
+
+class WaitlistViewModelTests: XCTestCase {
+
+    private let mockToken = "mock-token"
+    private let timestamp = 20
+
+    @MainActor
+    func testWhenNotificationServiceIsNilThenStatIsWaitlistRemoved() async {
```
Comment (samsymons) [inline LocalPackages/Waitlist/Tests/WaitlistTests/WaitlistViewModelTests.swift]: Tiny typo:

```suggestion
    func testWhenNotificationServiceIsNilThenStateIsWaitlistRemoved() async {
```

File: DuckDuckGo/MacWaitlistViewController.swift [hunk 3]
```diff
@@ -63,9 +64,21 @@ final class MacWaitlistViewController: UIViewController {
...
+        assertionFailure("Mac Waitlist is removed")
+    }
+
+    func waitlistViewModelDidOpenInviteCodeShareSheet(_ viewModel: WaitlistViewModel, inviteCode: String, senderFrame: CGRect) {
+        assertionFailure("Mac Waitlist is removed")
+    }
 
-    func macWaitlistViewModelDidOpenShareSheet(_ viewModel: MacWaitlistViewModel, senderFrame: CGRect) {
+    func waitlistViewModelDidOpenDownloadURLShareSheet(_ viewModel: WaitlistViewModel, senderFrame: CGRect) {
         let linkMetadata = MacWaitlistLinkMetadata()
```
Comment (samsymons) [inline DuckDuckGo/MacWaitlistViewController.swift:82]: Are we also able to remove the share sheet functionality + link metadata for the macOS waitlist? I think there's no longer any way to reach this state.

File: DuckDuckGo/WindowsBrowserWaitlistView.swift [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+                        .foregroundColor(.waitlistText)
+                        .padding(.top, 28)
+                        .padding(.bottom, 8)
+
+                    Text(UserText.windowsWaitlistInviteScreenStep1Description)
+                        .font(.proximaNova(size: 17))
+                        .foregroundColor(.waitlistText)
+                        .lineSpacing(6)
+
+                    if #available(iOS 14.0, *) {
```
Comment (samsymons) [inline DuckDuckGo/WindowsBrowserWaitlistView.swift]: We now have iOS 14 as our minimum deployment target, so the `else` branch here can go. 🎉

File: DuckDuckGo/WindowsBrowserWaitlistView.swift [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+                        InviteCodeView(title: UserText.waitlistInviteCode, inviteCode: inviteCode)
+                            .menuController(UserText.waitlistCopy) {
+                                action(.copyInviteCodeToPasteboard)
+                            }
+                            .padding(.top, 28)
```
Comment (samsymons) [inline DuckDuckGo/WindowsBrowserWaitlistView.swift]: I was still seeing sizing issues on the Menu Controller view, it's possible to somewhat work around them this way:

```suggestion
                        InviteCodeView(title: UserText.waitlistInviteCode, inviteCode: inviteCode)
                            .menuController(UserText.waitlistCopy) {
                                action(.copyInviteCodeToPasteboard)
                            }
                            .fixedSize()
                            .padding(.top, 28)
```

This approach will fix the sizing issue, but the view will no longer be vertically centered.

**Review by samsymons — APPROVED** [review]
LGTM, tested this end to end a number of times on iPad and iPhone – hard to find much to complain about here, really nicely done. 👏 The reusability changes here are really nice, and the feature flag support is looking great.

File: DuckDuckGo/MacWaitlistViewController.swift (hunk 3 above)
Comment (ayoy) [reply DuckDuckGo/MacWaitlistViewController.swift:82]: It actually needs to stay because we can still share macOS app download URL.

Comment (samsymons) [reply DuckDuckGo/MacWaitlistViewController.swift:82]: Of course 🤦‍♂️

**Pull request merged**